### ISSUE TITLE
GitAhead doesn't store credentials on Linux

### ISSUE THREAD

**Issue by funbiscuit** [gitahead/gitahead#204]
As was mentioned in another issue (closed) git ahead should use either gnome-keyring or libsecret as long as one of them is available, posted by @hackhaslam in https://github.com/gitahead/gitahead/issues/164#issuecomment-464284808_

I have both of them installed but still GitAhead asks each time for passprhase for SSH key (or for password if I use HTTPS connection). I have storage of credentials enabled in settings (it was on by default). So what should I do?

I tried to enable libsecret for git storage but that doesn't help:
git config --global credential.helper /usr/lib/git-core/git-credential-libsecret
git config --global credential.helper 'cache --timeout=3600'

I am on Manjaro Linux, installed packages are latest. I even installed lib32-libsecret just in case, but still the same.

**Comment by hackhaslam** [comment]
Credential helpers are not used to store SSH passphrases to unlock private keys. The best option for SSH is to set up ssh-agent to unlock your keys at login. GitAhead should automatically detect if ssh-agent is running.

To debug HTTPS password lookup, try running the bundled credential helpers from the command line. For example:

    cd path/to/gitahead
    ./git-credential-gnome-keyring get
    > protocol=https
    > host=github.com
    > username=<username>
    > ^D

If it works it should print out the stored password.

GitAhead doesn't actually respect the configured credential helper settings. On Linux it tries libsecret first, then gnome-keyring, then falls back on caching the password until GitAhead is closed.

**Comment by funbiscuit** [comment]
I'll try ssh-agent later but from what you're saying gitahead should at least cache password while it is running (and this would be okay for me). But it doesn't (neither SSH or HTTPS). Everytime I press fetch/push I need to enter passphrase/password (even if only one second passed). Is it how it should work?
I tried to lookup password how you said and after I press ^D on third line, I get third line duplicated:
[funbiscuit@funbiscuit-pc git-core]$ ./git-credential-libsecret get
protocol=https
host=github
username=funbiscuit
username=funbiscuit

And it does the same for any input so either I'm doing it wrong or I don't have password saved

**Comment by hackhaslam** [comment]
Yes, it should cache the password for HTTPS (not SSH). It probably isn't falling back on the cache implementation because it thinks that libsecret or gnome-keyring was successful.

Look at the gnome credentials interface (if you're using gnome) to see if you have any credentials associated with that host and username. If you do, then you can try to remove it to see if GitAhead will write a new one.

**Comment by funbiscuit** [comment]
I'm using KDE, not gnome.  And I installed gnome-keyring only after I got that problem with not storing credentials (for https) in hope that maybe it will help (it didn't). Now I uninstalled both gnome-keyring and lib32-libsecret but it still doesn't remember my password. And I can't uninstall libsecret (to check if caching works) because other packages require it.
I also tried to uncheck "store credentials in secure storage" in hope that now it will at least remember them while it is running. But still I need to type password every time I clone/push/anything

**Comment by funbiscuit** [comment]
I've enabled ssh-agent in the system but have a strange issue. After I login in system and try to clone something in terminal it asks for passphrase and after that works normally (i can fetch/push without typing passphrase again). I can even launch gitahead at this point and it will not ask for passphrase, I can fetch/push without it.
But if after logging into the system I launch gitahead immediately, it asks for passprhase. And keeps asking every time. Only when I fetch/push something in terminal and enter passphrase there, only after that gitahead stops asking for passphrase.

**Comment by hackhaslam** [comment]
It looks like there's an issue with the "store credentials ..." setting. GitAhead has to be restarted for it to take effect. Does it start caching after you restart? I will fix the issue so that it takes effect immediately.

I believe that KDE has its own credentials interface that will show libsecret credentials. It would be helpful to know if you can find anything there.

**Comment by hackhaslam** [comment]
It sounds like your ssh-agent may be running but you don't have any keys added to it. GitAhead doesn't know how to add keys yet. Does your ~/.ssh/config file contain a key like `AddKeysToAgent`?

**Comment by funbiscuit** [comment]
> GitAhead has to be restarted for it to take effect

Yes, after restart it asks only one time for ssh passphrase with the setting off. Well, it is already good enough for me if I only need to type it one time per app launch.

As for ssh, I installed blank manjaro in VM and gitahead. I configured ssh as in this answer:
https://stackoverflow.com/a/38980986
export of environment variable I've added to `~/.bashrc`
Now after I login I can open terminal and will need to enter passphrase one time when I clone/push. If I open any other terminal I don't need to enter it again. But with gitahead (with storing of credentials turned off) I need to enter passphrase each time I launch application (as I said I can launch new terminal and passphrase is not needed)
One thing is strange though. Even though I don't need to enter passphrase in new terminal for cloning, if I execute `ssh-add` I get following output:
```
Enter passphrase for /home/user/.ssh/id_rsa: 
Identity added: /home/user/.ssh/id_rsa
```
I can execute it multiple times (in one terminal or in multiple), each time everything the same. Shouldn't it say that key is already added?

**Comment by hackhaslam** [comment]
I fixed the "store credentials..." bug so that it applies immediately. That will be in the next build.

I'm not sure if I understand the SSH issues fully. I do think that GitAhead should be changed to support the `AddKeysToAgent` key if possible. That will at least make it work more like git.

**Comment by funbiscuit** [comment]
Thanks, that was not a big issue but still. It would be great if gitahead was supporting `AddKeysToAgent` but for me it doesn't matter a lot since I autolaunch it on login and keep open until logout. So I already need to enter pass only one time per login.
By the way, from all git guis for linux I've tried, this one is so far the best one for me. I've used gitkraken previously but it lags a lot. Not as much as responsive as it should be.
As a feature request - is it possible to highlight somehow which lines are for each part of blame heatmap? It is located sideways and it's hard to figure out at what line next portion of blame heatmap begins. For example it could be just horizontal lines dividing the document, or portion of blame heatmap could be highlighted when cursor is positioned on line, or even both. Should I create a new issue for this feature request? I've skimmed existing issues but haven't found one requesting such a feature.

**Comment by hackhaslam** [comment]
Yes, please add a new issue to track the blame feature request.

**Comment by erm3nda** [comment]
> That will be in the next build.

Has it been released? I see no mention about what's "this issue" build version, so i don't understand if i should still see that credential box every time i do open the program. It is perfect except for that annoying box.

**Comment by hackhaslam** [comment]
@erm3nda The current release is 2.5.8. That included a couple of potential fixes for issues related to storing/retrieving credentials on Linux. If you're still seeing an unexpected prompt for password after upgrading, then it would be useful to get some more details about your environment. What distro and window manager are you using? What protocol are you using to connect to the remote, etc.? You can also try to debug using the same steps I mentioned above.

**Comment by erm3nda** [comment]
@hackhaslam I'm using the latest release (2.5.8). GitAhead asks for keyring unloking, but GitAhead still asks me for gitlab after that.

My enviroment is Manjaro Linux, with kernel 4.19.60-1 and Deepin desktop.
My auth method is https with Token (I do use Token in the passwd field).
I dont know what's my Window Manager (kwin?)

The funny thing is that i can simply cancel the credential box, and my passwords are "loaded" because i can push with no problem. So, in my case, is not about not getting credentials, seems about a dummy box appearing for no reason. Fyi, credentials box appears empty.

**Comment by ysalmon** [comment]
I confirm problems with the handling of SSH keys with GitAhead 2.5.8 and Linux (I am using KUbuntu 18.04 but with gnome-keyring). When opening a repo in GitAhead after booting, I get prompted (by my agent), successively, for the passphrase of two different private keys, neither of which being the key for the corresponding domain as set out in ``~/.ssh/config``.

Of course if I run git fetch manually, I get prompted (once) for the passphrase of the correct key. Therafter, GitAhead does not prompt for a key anymore.

**Comment by hackhaslam** [comment]
@ysalmon Do you have the `AddKeysToAgent` key set in your `~/.ssh/config file`? If you do, then it probably starts working after fetching from the command line because GitAhead starts to get the key from the agent instead of trying to read it from disk.

**Comment by ysalmon** [comment]
Yes, I have it.

**Comment by ysalmon** [comment]
After some step-by-step debugging, it seems that GitAhead tries to initiate a SSH session to the remote with username ``git`` : there are calls to ``ssh_agent_auth`` where ``c->username`` is ``git``. Might this be the cause of the random passphrase prompt by the agent (my ssh config file for the relevant host explicitly states a username, and it is not ``git``, so the key rule would be disregarded by the agent).

**Comment by hackhaslam** [comment]
The username `git` comes from the remote URL. GitAhead does *not* try to reset the username to the one in your SSH config file. I don't know if it should or not.

@ysalmon in your case, is the issue that GitAhead prompts to decrypt the wrong key?

**Comment by ysalmon** [comment]
There are several parts of GitAhead that cause [``Remote::Callbacks::credentials``](https://github.com/gitahead/gitahead/blob/08727c527f57847ecef863f4ea3599495cfd911a/src/git/Remote.cpp#L204) to be called. Is several cases, the control flow goes to [``return git_cred_ssh_key_from_agent(out, name);``](https://github.com/gitahead/gitahead/blob/08727c527f57847ecef863f4ea3599495cfd911a/src/git/Remote.cpp#L236), then returns to various functions related to ssh and this makes the agent prompt for the passphrase of several, unrelated keys, and does not make it prompt for the right one.

Then other calls to [``Remote::Callbacks::credentials``](https://github.com/gitahead/gitahead/blob/08727c527f57847ecef863f4ea3599495cfd911a/src/git/Remote.cpp#L204) happen and this time the control flows to the [// Read SSH config file](https://github.com/gitahead/gitahead/blob/08727c527f57847ecef863f4ea3599495cfd911a/src/git/Remote.cpp#L242) part, which seems to find the correct key. GitAhead then prompts for a passphrase, which I assume is for the intended key (regretfully, the GitAhead prompt does not tell which key it is prompting for), but fetching fails with ``Unable to fetch from 'origin' - Failed to authenticate SSH session: Unable to send userauth-publickey request``.

**Comment by erm3nda** [comment]
I really expect to have gitahead working in the same way than it works in Windows, out of the box.
Im not getting if this has been fixed or everyone has the same problem than me.

I've changed to Manjaro gnome, the problem persist. 

I read somewhere in the webpage "professional grade software", but i can't figure out how something that works in Windows doesn't work in Linux. Should i change to work in Windows?

Don't get me wrong, i think that Gitahead is so cool, but those kind of annoying bug makes me think that i should change to any other git client.

Regards.

**Comment by erm3nda** [comment]
I've marked the option "store credentials in secure storage". Now it keeps asking me everytime that I open the repo. 
This is horrible guys.

I've ensure that both `gnome-keyring` and `libsecret` are installed.

**Comment by eduardoxfurtado** [comment]
I have set Automatic actions > Fetch every 1 minutes

Now I have a feeling that I'm asked the password for my open repos every minute.
Also, I can't know what repo is asking for the password, if it's my github or my bitbucket password, for example.

I have Credentials > Store credentials in secure storage > 1 (activated)


I'm using gitahead exactly because I want it to take care of my credentials with github and bitbucket, so they are integrated and I'm never asked them again to do fetches, pushes, pulls, clones, etc...

**Comment by erm3nda** [comment]
I've ended up using SSH keys :(

**Comment by eduardoxfurtado** [comment]
> SSH key

How can I set that up?
Why the sad emoji, it's not a perfect solution? Maybe the SSH keys are stored in plain text...

**Comment by erm3nda** [comment]
> How can I set that up?
https://docs.gitlab.com/ee/gitlab-basics/create-your-ssh-keys.html

There are a couples of ways, I used:
```
# Private GitLab instance
Host gitlab.org
  Preferredauthentications publickey
  IdentityFile ~/.ssh/gitlab_key_rsa
```


> Why the sad emoji, it's not a perfect solution? 
Sad emoji because I've been forced to do it. Although it's the recommended way, for me was good enough to simply fill username/token to login vía https.
> Maybe the SSH keys are stored in plain text...
Your username/password too, so it's not something you can point to. The advantage over the SSH keys is that is not easily sniffable for someone who looks over your shoulders, and you can save it into an usb, then configure the IdentityFile properly to point to it.

Regards.

**Comment by eduardoxfurtado** [comment]
Thank you @erm3nda .
@hackhaslam , do you think this feature to save the credentials are soon going to work properly?
I've been recommending gitahead to the people I know, but everyone seems to be having this same problem when they try it, that they are having to put their passwords for github, bitbucket, gitlab, etc, all the time

**Comment by hackhaslam** [comment]
It's hard to say when it will be fixed until I can reproduce it and pin down what the actual issue is.

**Comment by eduardoxfurtado** [comment]
Any way I can help with that?

I just tried lauching gitahead from the terminal to see if there was any output, but there was nothing there.

I tried to disable and re-enable the "Store credentials in secure storage".

Do I need to have some dependency installed for it to work?

I'm on Manjaro 18, with kernel 5.4.12, KDE plasma 5.17.5

**Comment by hackhaslam** [comment]
See my first response to this issue for steps to debug. You need either libsecret or gnome-keyring installed.

**Comment by eduardoxfurtado** [comment]
So, I just tried installing on a virtual machine Manjaro with Gnome (mine uses KDE), and it worked.
So my guess is that the problem has something to do with KDE.
I will try to install a fresh Manjaro with KDE on a virtual machine to see if the problem is present there.

**Comment by eduardoxfurtado** [comment]
@hackhaslam 
On the fresh install of the KDE installation the feature was not working.


After that I uninstalled gnome-keyring on my virtual machine where I had manjaro + gnome.
This caused the problem to start happening on this machine.

The next thing I did was installing gnome-keyring on the KDE virtual machine, which is not ideal. But with gnome-keyring installed, gitahead was storing the password!

This indicates that the problem is that gitahead is not being able to use libsecret to store the passwords, and only if gnome-keyring is there it works.

I think this is a lot of info to solve this bug now =)

to reproduce it, basically uninstall gnome-keyring

my libsecret version is 0.20.1-1

**Comment by eduardoxfurtado** [comment]
I'm touching the code now. Everything seems to be in order, so maybe it's a libsecret problem?
I can't figure out how to debug further =/

**Comment by Gr3q** [comment]
@hackhaslam I can give you some context why you can't reproduce the issue.

I'm running Arch Linux. If I build gitahead myself or I install gitahead from [AUR](https://aur.archlinux.org/packages/gitahead/) (what builds it for you), using credentials from libsecret works or from anywhere else works just fine.

This issue happens when you install gitahead via the [gitahead-bin](https://aur.archlinux.org/packages/gitahead-bin) package via AUR, what is using the the [prebuilt package](https://github.com/gitahead/gitahead/releases/download/v2.6.1/GitAhead-2.6.1.sh).

This could give you some pointers where the problem is.

**Comment by hackhaslam** [comment]
FWIW I tried a clean install of Manjaro Linux with with KDE. This has libsecret but no gnome-keyring. The error I get when I run the libsecret credential helper from the command line is:

    lookup failed: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.secrets was not provided by any .service files

I've found many similar issues, but the solution is always to install gnome-keyring. So libsecret is installed, but it doesn't really work without gnome-keyring? I don't really understand. It seems broken at some level. If someone has some insight that would help me to understand, I would appreciate it.

Of course, GitAhead could still give a better error message or even drop the libsecret credential helper entirely if it doesn't actually work without gnome-keyring.

**Comment by Gr3q** [comment]
@hackhaslam as far as I know libsecret is the command-line tool to access the gnome-keyring via DBus, I read it [here](https://wiki.archlinux.org/index.php/GNOME/Keyring#Installation) correct me if I'm wrong.

If that is right it makes sense that libsecret does not work without gnome-keyring (I have it installed thats why it works for me).

**Comment by hackhaslam** [comment]
@Gr3q I think that documentation is about secret-tool. The actual libsecret documentation is here: https://wiki.gnome.org/Projects/Libsecret. It explicitly states:

> libsecret replaces libgnome-keyring.

However, it also says:

> libsecret is a library for storing and retrieving passwords and other secrets. It communicates with the "Secret Service" using D-Bus. gnome-keyring and ksecretservice are both implementations of a Secret Service.

So maybe one of those things are required to actually store credentials. It still seems bizarre to me that these distos would have the libsecret package installed without its required dependencies. Seems broken.

**Comment by eduardoxfurtado** [comment]
so, maybe a propper solution would be changing the authentication sessions
code to use of something instead of libsecret, or to include gnome-keyring
in the requirements of gitahead.

the latter I do not find too elegant, because it's too much dependant on a
platform not used by majority? idk

**Comment by hackhaslam** [comment]
GitAhead includes the libsecret and gnome-keyring credential helpers from git. It tries libsecret first since it was meant to be the new replacement for gnome-keyring (at least that's what I thought). If it can't find libsecret (because it can't link to libsecret.so) it falls back on gnome-keyring. If it can't find gnome-keyring it falls back on caching the credentials in memory during the current session.

In this broken scenario, libsecret is installed, but it can't actually store or look up credentials. GitAhead could be fixed detect when libsecret fails in this way and still try to fall back on gnome-keyring. Of course that won't really fix anything in this case since gnome-keyring isn't installed. It would just allow it to gracefully fall back on the caching strategy.

**Comment by ysalmon** [comment]
Hello,

thanks for this discussion.

What is not clear to me is why Githead needs to ship libsecret or
whatever and why it needs to make a decision about what it uses first.

When I use git from the command line, it uses whatever agent is active
(I think it relies on the SSH_AUTH_SOCK environment variable) ; I do not
remember setting git to use seahorse. Would this not be a good "first
try" for GitAhead ?

Or is there some behind the scene autoconfiguration that I am not aware of ?

**Comment by Gr3q** [comment]
They are optional dependencies at best so they should be identified as such (in packages and also in readme at first).

@hackhaslam KDE uses kwallet at the moment because ksecretservice is not ready yet (for years, actually), so libsecret is installed for nothing. libsecret is only for accessing the gnome-keyring at the moment.

The actual flow for checking should look like this checking for keyrings:
![flow](https://user-images.githubusercontent.com/22377521/75147592-941b1300-56f5-11ea-9f1e-9e35b8faeea6.png)

**Comment by Gr3q** [comment]
In addition, we could check the users global git config first if any credential-helper is set, and try to use that first.

**Comment by eduardoxfurtado** [comment]
the problem with that is that gitahead does not work with libahead and
without gnome keyring. To me, this indicates that caching the session on
the system is not working properly.

when gnome keyring is present it works, this would indicate that probably
libsecret doesn't work or it doesn't do what who wrote the code thought it
did.

anyway, I still think that gnome-keyring is not an elegant solution, maybe
there is another lib that can replace libahead on linux or even what is
being used in a windows or OSX context

**Comment by hackhaslam** [comment]
@Gr3q Yes, I agree that GitAhead should check the git config.

@eduardoxfurtado GitAhead uses the Windows Credential Manager on Windows and Keychain on macOS. I don't think that it would be appropriate to roll our own secure credential storage solution on Linux. It's too potentially problematic.

**Comment by sonicdebris** [comment]
Hello in case someone like me is still having issues on linux and like me finds this thread, I had installed this on Manjaro through AUR and after going to `opt/gitahead` and making `./git-credential-libsecret` executable (it was not), credential storing started working.

**Comment by Gr3q** [comment]
@sonicdebris can you write down the same thing as a new comment for gitahead-bin package on AUR?

This is a packaging issue, the PKGBUILD is responsible setting permissions on install.

**Comment by eduardoxfurtado** [comment]
It didn't work here :(
I am using Manjaro too, with gitahead installed from AUR.
These are the two files that I can find on my system:

-rwxr-xr-x 1 root root 18K Jul 16 11:04
/usr/share/gitahead/git-credential-libsecret*
-rwxr-xr-x 1 root root 19K Jun  1 15:41
/usr/lib/git-core/git-credential-libsecret*

**Comment by ginanck** [comment]
Hi Guys,

I'am using 20.6.3 version of GitAhead on Kubuntu 20.04

And still it can not store the passwords. It was working before.

I've enabled 2 factor verification for Github and Bitbucket, then it can not store my password anymore

And always asking when pull, push request including after complete the requests asking again

Could you help about it :)

**Comment by jbsilva** [comment]
Same problem here. Only worked after installing `gnome-keyring`.
